panic: encode struct of single map-typed field

- code
```
type Wrap struct {
	SingleMap *SingleMap `bson:"single_map,omitempty" json:"single_map,omitempty"`
}

type SingleMap struct {
	//Any      string `json:"any"`
	M map[string]string `bson:"m" json:"m"`
}

func TestMarshal_SingleMap(t *testing.T) {
	var data = `[{
		"single_map": {
			"m": {
				"a": "b"
			}
		}
	}]`
	var obj = new([]*Wrap)
	if err := json.Unmarshal([]byte(data), obj); err != nil {
		t.Fatal(err)
	}
	fmt.Printf("%#v\n", obj)
	
	buf, err := Encode(obj, 0)
	if err != nil {
		t.Fatal(err)
	}
	fmt.Println(string(buf))
}
```
- result
```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1011f6b]

goroutine 6 [running]:
testing.tRunner.func1.2(0x1417e60, 0x170de30)
	/usr/local/go/src/testing/testing.go:1143 +0x332
testing.tRunner.func1(0xc000001b00)
	/usr/local/go/src/testing/testing.go:1146 +0x4b6
panic(0x1417e60, 0x170de30)
	/usr/local/go/src/runtime/panic.go:965 +0x1b9
github.com/bytedance/sonic/encoder.iteratorStart(0x1416320, 0xc000010508, 0x0, 0xc00013fcb0, 0x0, 0xc000300000)
	/Users/admin/Desktop/kitex/sonic3/encoder/mapiter.go:162 +0x16f
github.com/bytedance/sonic/encoder.encodeTypedPointer(0xc00000f500, 0x14209e0, 0xc00013fcb0, 0xc000400000, 0x0, 0xc000300000, 0xc000010508)
	/Users/admin/Desktop/kitex/sonic3/encoder/primitives.go:81 +0x447
github.com/bytedance/sonic/encoder.encodeTypedPointer(0xc00000f500, 0x13f6a40, 0xc000610008, 0xc000400000, 0x0, 0x0, 0xc00013fe60)
	/Users/admin/Desktop/kitex/sonic3/encoder/primitives.go:81 +0x447
github.com/bytedance/sonic/encoder.EncodeInto(0xc00000f500, 0x13f6a40, 0xc00000f3f8, 0x0, 0x30, 0x0)
	/Users/admin/Desktop/kitex/sonic3/encoder/encoder.go:101 +0xd4
github.com/bytedance/sonic/encoder.Encode(0x13f6a40, 0xc00000f3f8, 0x0, 0x4, 0xc00013ff50, 0x1, 0x1, 0x30)
	/Users/admin/Desktop/kitex/sonic3/encoder/encoder.go:79 +0xc5
github.com/bytedance/sonic/encoder.TestMarshal_SingleMap(0xc000001b00)
	/Users/admin/Desktop/kitex/sonic3/encoder/issue40_test.go:230 +0x19e
testing.tRunner(0xc000001b00, 0x14978b0)
	/usr/local/go/src/testing/testing.go:1193 +0xef
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:1238 +0x2b3
exit status 2
```

--- a/encoder/assembler_amd64.go
+++ b/encoder/assembler_amd64.go
@@ -1022,7 +1022,7 @@ func (self *_Assembler) _asm_OP_recurse(p *_Instr) {
     self.Emit("MOVQ", _AX, jit.Ptr(_SP, 8))     // MOVQ AX, 8(SP)
 
     /* check for indirection */
-    if p.vk() == reflect.Ptr {
+    if (p.vf() & rt.F_direct) != 0 {
         self.Emit("MOVQ", _SP_p, _AX)               // MOVQ SP.p, AX
     } else {
         self.Emit("MOVQ", _SP_p, jit.Ptr(_SP, 48))  // MOVQ SP.p, 48(SP)

--- a/encoder/compiler.go
+++ b/encoder/compiler.go
@@ -224,6 +224,10 @@ func (self _Instr) vi() int {
     return rt.UnpackInt(self.u)
 }
 
+func (self _Instr) vf() uint8 {
+    return (*rt.GoType)(self.p).KindFlags
+}
+
 func (self _Instr) vs() (v string) {
     (*rt.GoString)(unsafe.Pointer(&v)).Ptr = self.p
     (*rt.GoString)(unsafe.Pointer(&v)).Len = self.vi()

--- a/issue76_test.go
+++ b/issue76_test.go
@@ -1,0 +1,51 @@
+/*
+ * Copyright 2021 ByteDance Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package sonic
+
+import (
+    `encoding/json`
+    `fmt`
+    `testing`
+
+    `github.com/davecgh/go-spew/spew`
+)
+
+type SingleMapField struct {
+    Z *int
+}
+
+type SingleMapFieldOuter struct {
+    Y *SingleMapField
+}
+
+type SingleMapFieldOuterContainer struct {
+    X *SingleMapFieldOuter
+}
+
+func TestIssue76_MarshalSingleMapField(t *testing.T) {
+    data := `{"X": {"Y": {"Z": 1}}}`
+    obj := new(SingleMapFieldOuterContainer)
+    if err := json.Unmarshal([]byte(data), obj); err != nil {
+        t.Fatal(err)
+    }
+    spew.Dump(obj)
+    buf, err := Marshal(obj)
+    if err != nil {
+        t.Fatal(err)
+    }
+    fmt.Println(string(buf))
+}

